toml.abnf

```diff
@@ -84,9 +84,11 @@ ml-basic-string = ml-basic-string-delim ml-basic-body ml-basic-string-delim
 
 ml-basic-string-delim = 3quotation-mark
 
-ml-basic-body = *( ml-basic-char / newline / ( escape ws newline ) )
-ml-basic-char = ml-basic-unescaped / escaped
-ml-basic-unescaped = wschar / %x21-5B / %x5D-7E / non-ascii
+ml-basic-body = *( ml-basic-char / wschar / newline / ml-basic-escaped-ws-nl )
+ml-basic-char = ( ml-basic-unescaped / escaped )
+ml-basic-unescaped = %x21-5B / %x5D-7E / non-ascii
+ml-basic-nonblank = %x21-7E / non-ascii
+ml-basic-escaped-ws-nl = escape newline *( wschar / newline ) ml-basic-nonblank
```

Comment:
I don't understand what's `ml-basic-nonblank` token is doing at the end of this rule. Supposedly, that non-blank character should NOT be escaped, but rather be part of the normal string content? Also, it's possible that the "line ending backslash" + subsequent ignored whitespace is immediately followed by a "closing delimiter", but the  `ml-basic-nonblank` token prevents that.

My impression is that the `ml-basic-nonblank` token could, and should, be removed to get the correct behavior. That's true, at least, if ABNF rules are "greedy", matching as many `wschar / newline` as the end of the `ml-basic-escaped-ws-nl` as they possibly can – which in my understanding is the case. ('The [ABNF] repetition operator is "greedy", meaning that it always matches the longest possible string. It never backtracks to see if a shorter match would succeed', says [the Internet](http://www.coasttocoastresearch.com/overview).)

Reply:
Also, I'd propose to rename the rule to `ml-basic-escaped-nl`. After all, the escaped thing is always the newline, not any other whitespace (though other whitespace might follow the escaped newline).

Reply:
The fact that ABNF operator `*` is greedy solves everything!

That is not specified in ABNF RFC but RFC stands for "Request For Comments", not "Rules Forever Constant".  What matters is that the message reach parser implementers.

I see reason to have the `ml-basic-escaped-nl` at the end, and we can now have semantics back.

I amended the suggested changes to the commit.

Reply:
To my greatest surprise, instaparse is non-greedy!  (noticed that while checking).

https://github.com/Engelberg/instaparse#regular-expressions-a-word-of-warning

Do we still consider ABNF's "*" to be greedy?

Reply:
It is not either greedy or non-greedy, it shows all alternative forms.

With the original version, we have only one possible form (because it only takes the first space).

What a storm of irony: I created the problem I was trying to solve!